File: 110-balanced-binary-tree/balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
//  Test upload
class Solution {
public:
    int dfs(TreeNode *node){
        if(node == NULL){
            return 0;             //a null node is balanced with a height of 0
        }

        return max(dfs(node->left), dfs(node->right)) + 1;
    }
    
    bool isBalanced(TreeNode* root) {         // kinda like a top down approach
        if(root == NULL){
            return true;
        }

        int left_height = dfs(root->left);
        int right_height = dfs(root->right);

        if(abs(left_height - right_height) > 1){     // if we find/hit any unbalanced node => exit to false
            return false;
        }

        return isBalanced(root->left) && isBalanced(root->right);
    }
};

File: 110-balanced-binary-tree/balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def dfs(node):
            if node == None:
                return [True, 0]  # a null node is balanced and has a height of 0
            
            left, right = dfs(node.left), dfs(node.right)
            
            balanced = left[0] and right[0] and abs(left[1] - right[1]) <= 1
            height = max(left[1], right[1]) + 1

            return [balanced, height]

        return dfs(root)[0]
        